pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1030,31 +1109,26 @@ func (r *ReconcileClusterDeployment) syncDeletedClusterDeployment(cd *hivev1.Clu
 		return reconcile.Result{}, err
 	}
 
-	// Delete the install job in case it's still running:
-	installJob := &batchv1.Job{}
-	err = r.Get(context.Background(),
-		types.NamespacedName{
-			Name:      install.GetInstallJobName(cd),
-			Namespace: cd.Namespace,
-		},
-		installJob)
-	if err != nil && apierrors.IsNotFound(err) {
-		cdLog.Debug("install job no longer exists, nothing to cleanup")
-	} else if err != nil {
-		cdLog.WithError(err).Errorf("error getting existing install job for deleted cluster deployment")
-		return reconcile.Result{}, err
-	} else if !installJob.DeletionTimestamp.IsZero() {
-		cdLog.WithField("finalizers", installJob.Finalizers).Info("install job is being deleted, requeueing to wait for deletion")
-		return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
-	} else {
-		err = r.Delete(context.Background(), installJob,
-			client.PropagationPolicy(metav1.DeletePropagationForeground))
-		if err != nil {
-			cdLog.WithError(err).Errorf("error deleting existing install job for deleted cluster deployment")
+	// Wait for outstanding provision to be removed before creating deprovision request
+	if cd.Status.Provision != nil {
+		provision := &hivev1.ClusterProvision{}
+		switch err := r.Get(context.TODO(), types.NamespacedName{Name: cd.Status.Provision.Name, Namespace: cd.Namespace}, provision); {
+		case apierrors.IsNotFound(err):
+			cdLog.Debug("linked provision removed")
+		case err != nil:
+			cdLog.WithError(err).Error("could not get provision")
 			return reconcile.Result{}, err
+		case provision.DeletionTimestamp == nil:
+			if err := r.Delete(context.TODO(), provision); err != nil {
+				cdLog.WithError(err).Error("could not delete provision")
+				return reconcile.Result{}, err
+			}
+			cdLog.Info("deleted outstanding provision")
+			return reconcile.Result{}, nil
+		default:
+			cdLog.Debug("still waiting for outstanding provision to be removed")
+			return reconcile.Result{}, nil
```

Comment:
I don't think so. My understanding is that the clusterdeployment will be requeued when the clusterprovision is removed. It seems to be working that way, but I can dig into it deeper to verify that that is correct.

Reply:
My understanding is that you get an event when the delete action happens, but not when the resource actually goes away.

Reply:
OK. I will verify the behavior and add a requeue if needed.

Reply:
You are correct about the behavior. The ClusterDeployment is not requeued when the ClusterProvision is finally removed. It was working in my tests because the ClusterDeployment was being requeued as a result of other changes. I will add a RequeueAfter here.